htmresearch/algorithms/column_pooler.py

```diff
@@ -174,16 +192,127 @@ def compute(self, feedforwardInput=(), lateralInputs=(),
+        self._computeOnlineLearningMode(predictedActiveInput, lateralInputs,
+                                        feedforwardGrowthCandidates)
+      elif (numpy.abs(len(self.activeCells) - self.sdrSize)
+          > (self.learningTolerance*self.sdrSize)):
+        # If the pooler doesn't have a single representation, try to infer one,
+        # before actually attempting to learn.
+        self._computeInferenceMode(feedforwardInput, lateralInputs)
+        self._computeOnlineLearningMode(feedforwardInput, lateralInputs,
+                                        feedforwardGrowthCandidates)
+      else:
+        # If there isn't predicted input and we have a single SDR,
+        # we are extending that representation and should just learn.
+        self._computeOnlineLearningMode(feedforwardInput, lateralInputs,
+                                  feedforwardGrowthCandidates)
+        self._computeInferenceMode(feedforwardInput, lateralInputs)
+
+
+  def _computeOnlineLearningMode(self, feedforwardInput, lateralInputs,
+                                 feedforwardGrowthCandidates):
+    """
+    Learning mode: we are learning a new object in an online fashion. If there
+    is no prior activity, we randomly activate 'sdrSize' cells and create
+    connections to incoming input. If there was prior activity, we maintain it.
+    If we have a union, we simply do not learn at all.
+
+    These cells will represent the object and learn distal connections to each
+    other and to lateral cortical columns.
+
+    Parameters:
+    ----------------------------
+    @param  feedforwardInput (sequence)
+            Sorted indices of active feedforward input bits
+
+    @param  lateralInputs (list of sequences)
+            For each lateral layer, a list of sorted indices of active lateral
+            input bits
+
+    @param  feedforwardGrowthCandidates (sequence or None)
+            Sorted indices of feedforward input bits that the active cells may
+            grow new synapses to.  This is assumed to be the predicted active
+            cells of the input layer.
+    """
+    prevActiveCells = self.activeCells
+    cellsToLearn = set(self.activeCells)
+
+    # If there are not enough previously active cells, select random subset of
+    # cells to learn on.  If there were only some active cells, we assume that
+    # we are learning an object related to what is represented by those cells;
+    # if there were none, we are learning a new object.
+    # This case is the only way different object representations are created.
+    if len(cellsToLearn) < self.sdrSize:
+      if len(cellsToLearn) < self.sdrSize*(1 - self.learningTolerance):
+        cellsToLearn = set()
+      numToAdd = self.sdrSize - len(cellsToLearn)
+      newCells = _sampleRange(self._random,
+                              0, self.numberOfCells(),
+                              step=1, k=numToAdd)
+      cellsToLearn |= set(newCells)
```

Comment:
It would be best to do this all with numpy arrays. Sets are slower because every element of the set can be of any type, while every element of a numpy array is of a single type, so the underlying C code can work very quickly.

~~~python
cellsToLearn = np.union1d(cellsToLearn, newCells)
~~~